diff --git a/src/main/java/org/apache/karaf/tooling/features/GenerateDescriptorMojo.java b/src/main/java/org/apache/karaf/tooling/features/GenerateDescriptorMojo.java
index 65d4aa8..ec351b4 100644
--- a/src/main/java/org/apache/karaf/tooling/features/GenerateDescriptorMojo.java
+++ b/src/main/java/org/apache/karaf/tooling/features/GenerateDescriptorMojo.java
@@ -74,7 +74,6 @@ import static org.apache.karaf.deployer.kar.KarArtifactInstaller.FEATURE_CLASSIF
  * @description Generates the features XML file starting with an optional source feature.xml and adding
  * project dependencies as bundles and feature/car dependencies
  */
-@SuppressWarnings("unchecked")
 public class GenerateDescriptorMojo extends AbstractLogEnabled implements Mojo {
 
     /**
